SG14/ring.h

```diff
@@ -157,26 +158,27 @@ namespace sg14
 // Sample implementation
 
 template <typename T>
-void sg14::null_popper<T>::operator()(T&)
+void sg14::null_popper<T>::operator()(T&) const noexcept
 {}
 
 template <typename T>
-T sg14::default_popper<T>::operator()(T& t)
+T sg14::default_popper<T>::operator()(T& t) const
 {
 	return std::move(t);
 }
 
 template <typename T>
-sg14::copy_popper<T>::copy_popper(T&& t)
-	: copy(std::move(t))
+sg14::copy_popper<T>::copy_popper(T t)
+	: m_copy(std::move(t))
 {}
 
 template <typename T>
-T sg14::copy_popper<T>::operator()(T& t)
+T sg14::copy_popper<T>::operator()(T& t) const
 {
-	T old = t;
-	t = copy;
-	return t;
+	T old = m_copy;
+	using std::swap;
+	swap(old, t);
+	return old;
```

Comment:
This dance makes the operation exception-safe, in case that matters. If copying `m_copy` into `old` throws an exception, then the swap doesn't happen and the value of `t` is preserved.